### PR TITLE
Implement --compare-dest & --copy-dest

### PR DESCRIPTION
**Pull request by yparitcher**
#### What is the purpose of this change?

Implement rsync's --compare-dest & --copy-dest flags to complement the existing --backup-dir to help with incremental backup strategies.

#### Was the change discussed in an issue or in the forum before?

Discussed in #3278 

#### Checklist

This depends on my other PR #3272 to work with --backup-dir

- [X] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: fs/operations/operations_test.go [hunk 1]
```diff
@@ -847,6 +847,86 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+		fs.Config.CompareDest = ""
+	}()
+
+	file2 := r.WriteObject(context.Background(), "CompareDest/one", "one", t1)
+	file1 := r.WriteFile("one", "one", t1)
+
+	fstest.CheckItems(t, r.Fremote, file2)
+	fstest.CheckItems(t, r.Flocal, file1)
+
+	fdst, err := fs.NewFs(r.FremoteName + "/dst")
```
Comment (golangcibot) [inline fs/operations/operations_test.go]: ineffectual assignment to `err` (from `ineffassign`)

File: fs/operations/operations_test.go [hunk 2]
```diff
@@ -847,6 +847,86 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+
+	err = operations.CopyFile(context.Background(), fdst, r.Flocal, file1.Path, file1.Path)
```
Comment (golangcibot) [inline fs/operations/operations_test.go]: ineffectual assignment to `err` (from `ineffassign`)

File: fs/operations/operations_test.go [hunk 3]
```diff
@@ -847,6 +847,86 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+	file2 := r.WriteObject(context.Background(), "CopyDest/one", "one", t1)
+	file1 := r.WriteFile("one", "one", t1)
+
+	fstest.CheckItems(t, r.Fremote, file2)
+	fstest.CheckItems(t, r.Flocal, file1)
+
+	fdst, err := fs.NewFs(r.FremoteName + "/dst")
+	require.NoError(t, err)
+
+	err = operations.CopyFile(context.Background(), fdst, r.Flocal, file1.Path, file1.Path)
```
Comment (golangcibot) [inline fs/operations/operations_test.go]: ineffectual assignment to `err` (from `ineffassign`)

File: fs/operations/operations.go [hunk 4]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+	var remote string
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CopyDestFile, _ := CopyDest.NewObject(ctx, remote)
+	if equal(ctx, src, CopyDestFile, fs.Config.SizeOnly, fs.Config.CheckSum, false) {
+		if dst == nil || !Equal(ctx, src, dst) {
+//			FIXME BackupDir depends on #3272
```
Comment (golangcibot) [inline fs/operations/operations.go]: File is not `goimports`-ed (from `goimports`)

File: fs/operations/operations.go [hunk 5]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+	var remote string
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CopyDestFile, _ := CopyDest.NewObject(ctx, remote)
+	if equal(ctx, src, CopyDestFile, fs.Config.SizeOnly, fs.Config.CheckSum, false) {
+		if dst == nil || !Equal(ctx, src, dst) {
+//			BackupDir depends on #3272
```
Comment (golangcibot) [inline fs/operations/operations.go]: File is not `goimports`-ed (from `goimports`)

File: fs/operations/operations.go [hunk 6]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CopyDestFile, _ := CopyDest.NewObject(ctx, remote)
+	if equal(ctx, src, CopyDestFile, fs.Config.SizeOnly, fs.Config.CheckSum, false) {
+		if dst == nil || !Equal(ctx, src, dst) {
+			// FIXME BackupDir depends on #3272
+//			if dst != nil && Backupdir != nil {
```
Comment (golangcibot) [inline fs/operations/operations.go]: File is not `goimports`-ed (from `goimports`)

File: fs/operations/operations.go [hunk 7]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+//
+// Returns True if src was copied from --copy-dest
+func CopyDest(ctx context.Context, fdst fs.Fs, dst, src fs.Object, CopyDest fs.Fs) bool {
+	var remote string
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CopyDestFile, _ := CopyDest.NewObject(ctx, remote)
```
Comment (ncw) [inline fs/operations/operations.go]: Same observation here about `CopyDestFile` being `nil`

File: fs/operations/operations.go [hunk 8]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+	if equal(ctx, src, CopyDestFile, fs.Config.SizeOnly, fs.Config.CheckSum, false) {
+		if dst == nil || !Equal(ctx, src, dst) {
+			// FIXME BackupDir depends on #3272
```
Comment (ncw) [inline fs/operations/operations.go]: #3272 is now merged so you can fix this up I think?

File: fs/operations/operations.go [hunk 9]
```diff
@@ -1519,7 +1600,23 @@ func moveOrCopyFile(ctx context.Context, fdst fs.Fs, fsrc fs.Fs, dstFileName str
...
 
-	if NeedTransfer(ctx, dstObj, srcObj) {
+	NoNeedTransfer := false
+	if fs.Config.CompareDest != "" {
+		CompareDestDir, err := GetCompareDest()
+		if err != nil {
+			return err
+		}
+		NoNeedTransfer = CompareDest(ctx, dstObj, srcObj, CompareDestDir)
+	}
```
Comment (ncw) [inline fs/operations/operations.go]: This should probably be `} else if {` since `CompareDest` and `CopyDest` are mutually exclusive

File: fs/operations/operations.go [hunk 10]
```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+			//	err = MoveBackupDir(ctx, backupDir, dst)
+			//	if err != nil {
+			//		return errors.Wrap(err, "moving to --backup-dir failed")
+			//	}
+			//	// If successful zero out the dstObj as it is no longer there
+			//	dst = nil
+			//}
+			_, err := Copy(ctx, fdst, dst, remote, CopyDestFile)
+			if err != nil {
+				fs.Debugf(src, "Destination found in --copy-dest, error copying")
```
Comment (ncw) [inline fs/operations/operations.go]: This should be an `Errorf` log.

Not bubbling up the error means that the sync won't fail and won't be retries if this fails.

I think that the sync should fail if the copy doesn't work - the user explicitly asked for it after all.

So I think this function needs to return an `error`

File: fs/operations/operations_test.go [hunk 11]
```diff
@@ -847,6 +847,95 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+func TestCopyFileCompareDest(t *testing.T) {
+	r := fstest.NewRun(t)
+	defer r.Finalise()
+
+	fs.Config.CompareDest = r.FremoteName + "/CompareDest"
+	defer func() {
+		fs.Config.CompareDest = ""
+	}()
+
+	file2 := r.WriteObject(context.Background(), "CompareDest/one", "one", t1)
```
Comment (ncw) [inline fs/operations/operations_test.go]: These tests didn't discover the `nil` pointer bug caused by not checking the error from `NewObject` which means that they need more in them.

I think you need another file where the file doesn't exist in `CompareDest`.

I recommend using a coverage tool so you can see you hit all the branches (or as many as you can!) in the code. Here is how to do it: https://blog.golang.org/cover

File: fs/operations/operations_test.go [hunk 12]
```diff
@@ -847,6 +847,95 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+	fstest.CheckItems(t, r.Flocal, file1b)
+
+	err = operations.CopyFile(context.Background(), fdst, r.Flocal, file1.Path, file1.Path)
+	require.NoError(t, err)
+
+	fstest.CheckItems(t, r.Fremote, file2b, file3)
+}
+
+// Test with CopyDest set
+func TestCopyFileCopyDest(t *testing.T) {
```
Comment (ncw) [inline fs/operations/operations_test.go]: Same comment as above.

File: fs/sync/sync.go [hunk 13]
```diff
@@ -61,6 +61,8 @@ type syncCopyMove struct {
 	trackRenamesWg sync.WaitGroup         // wg for background track renames
 	trackRenamesCh chan fs.Object         // objects are pumped in here
 	renameCheck    []fs.Object            // accumulate files to check for rename here
+	CompareDest    fs.Fs                  // place to check for files
```
Comment (ncw) [inline fs/sync/sync.go]: These should be small initial letter so they aren't exported, so `compareDest`, `copyDest`

File: fs/sync/sync.go [hunk 14]
```diff
@@ -219,7 +235,14 @@ func (s *syncCopyMove) pairChecker(in *pipe, out *pipe, wg *sync.WaitGroup) {
 		accounting.Stats.Checking(src.Remote())
 		// Check to see if can store this
 		if src.Storable() {
-			if operations.NeedTransfer(s.ctx, pair.Dst, pair.Src) {
+			NoNeedTransfer := false
+			if fs.Config.CompareDest != "" {
```
Comment (ncw) [inline fs/sync/sync.go]: I think this little block of code is duplicated 3 times, once in operations (with a few more things) and twice in sync.

How about factoring it into a function? Say `CompareOrCopyDest` returning a bool and an error most likely?

This would then let us make `CompareDest` and `CopyDest` private in `operations`

File: fs/sync/sync_test.go [hunk 15]
```diff
@@ -1202,6 +1202,93 @@ func TestSyncOverlap(t *testing.T) {
...
+func TestSyncCompareDest(t *testing.T) {
+	r := fstest.NewRun(t)
+	defer r.Finalise()
+
+	fs.Config.CompareDest = r.FremoteName + "/CompareDest"
+	defer func() {
+		fs.Config.CompareDest = ""
+	}()
+
+	file2 := r.WriteObject(context.Background(), "CompareDest/one", "one", t1)
```
Comment (ncw) [inline fs/sync/sync_test.go]: I think this probably needs more cases in it for the destination object
- destination object exists in destination dir
- destination object exists in CompareDest
- destination object exists in dir and CompareDest
- destination object does not exist

File: fs/sync/sync_test.go [hunk 16]
```diff
@@ -1202,6 +1202,93 @@ func TestSyncOverlap(t *testing.T) {
...
+
+	accounting.Stats.ResetCounters()
+	err = Sync(context.Background(), fdst, r.Flocal, false)
+	require.NoError(t, err)
+
+	fstest.CheckItems(t, r.Fremote, file2b, file3)
+}
+
+// Test with CopyDest set
+func TestSyncCopyDest(t *testing.T) {
```
Comment (ncw) [inline fs/sync/sync_test.go]: See above

File: fs/sync/sync.go [hunk 17]
```diff
@@ -142,6 +144,20 @@ func newSyncCopyMove(ctx context.Context, fdst, fsrc fs.Fs, deleteMode fs.Delete
 		}
 		s.suffix = fs.Config.Suffix
 	}
+	if fs.Config.CompareDest != "" {
+		var err error
+		s.CompareDest, err = operations.GetCompareDest()
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (ncw) [inline fs/sync/sync.go]: `} else if {`

File: fs/sync/sync.go [hunk 18]
```diff
@@ -219,7 +235,14 @@ func (s *syncCopyMove) pairChecker(in *pipe, out *pipe, wg *sync.WaitGroup) {
 		accounting.Stats.Checking(src.Remote())
 		// Check to see if can store this
 		if src.Storable() {
-			if operations.NeedTransfer(s.ctx, pair.Dst, pair.Src) {
+			NoNeedTransfer := false
+			if fs.Config.CompareDest != "" {
+				NoNeedTransfer = operations.CompareDest(s.ctx, pair.Dst, pair.Src, s.CompareDest)
+			}
```
Comment (ncw) [inline fs/sync/sync.go]: `} else if {`

**Comment by yparitcher** [comment]
sorry for the sloppy code, i am not so used to the Go syntax

File: fs/operations/operations.go [hunk 19]
```diff
@@ -110,7 +110,7 @@ func sizeDiffers(src, dst fs.ObjectInfo) bool {
 
 var checksumWarning sync.Once
 
-func equal(ctx context.Context, src fs.ObjectInfo, dst fs.Object, sizeOnly, checkSum bool) bool {
+func equal(ctx context.Context, src fs.ObjectInfo, dst fs.Object, sizeOnly, checkSum, UpdateModTime bool) bool {
```
[the remark this answers is not in the document]
Comment (yparitcher) [reply fs/operations/operations.go]: the existing equal function updates the modtime unless the `NoUpdateModTime` flag is passed, when comparing to the `compare-dest` or `copy-dest` i do not want to update the mod time a they are presumed to be an older state which should not be modified.
i thought that changing the signature would be cleaner than hacking into the `NoUpdateModTime` config, especially as this is a local function that is only used by the `Equal` convenience function.

File: fs/operations/operations.go (hunk 10 above)
Comment (yparitcher) [reply fs/operations/operations.go]: good point

File: fs/sync/sync.go (hunk 14 above)
Comment (yparitcher) [reply fs/sync/sync.go]: will do

**Comment by ncw** [comment]
> sorry for the sloppy code, i am not so used to the Go syntax

No apologies necessary - I think you are doing very well and I really appreciate your submissions :-)

File: fs/operations/operations_test.go (hunk 11 above)
Comment (yparitcher) [reply fs/operations/operations_test.go]: thanks for the link, will write better coverage

**Comment by yparitcher** [comment]
refactored, wrote better tests & found a few corner cases.
please review.

**Comment by ncw** [comment]
Just kicked the CI (yet another flakey test!) will review in a bit

File: fs/operations/operations.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ncw) [reply fs/operations/operations.go]: Yes that makes sense - thanks for explaining :-)

File: fs/operations/operations_test.go [hunk 20]
```diff
@@ -866,6 +866,183 @@ func TestCopyFileBackupDir(t *testing.T) {
...
+
+	fstest.CheckItems(t, r.Fremote, file2, file3, file4, file5bdst)
+}
+
+// Test with CopyDest set
+func TestCopyFileCopyDest(t *testing.T) {
+	r := fstest.NewRun(t)
+	defer r.Finalise()
+
+	if !operations.CanServerSideMove(r.Fremote) {
```
Comment (ncw) [inline fs/operations/operations_test.go]: Server side copy is necessary for this isn't it?

File: fs/operations/operations.go [hunk 21]
```diff
@@ -1360,6 +1360,115 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+// GetCopyDest sets up --copy-dest
+func GetCopyDest(fdst fs.Fs) (CopyDest fs.Fs, err error) {
+	CopyDest, err = cache.Get(fs.Config.CopyDest)
+	if err != nil {
+		return nil, fserrors.FatalError(errors.Errorf("Failed to make fs for --copy-dest %q: %v", fs.Config.CopyDest, err))
+	}
+	if !SameConfig(fdst, CopyDest) {
+		return nil, fserrors.FatalError(errors.New("parameter to --copy-dest has to be on the same remote as destination"))
+	}
+	if !CanServerSideMove(CopyDest) {
```
Comment (ncw) [inline fs/operations/operations.go]: I think this should be checking for server side copy only. This doesn't move the files it copies them doesn't it?

File: fs/operations/operations.go [hunk 22]
```diff
@@ -1360,6 +1360,115 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
...
+		return nil, fserrors.FatalError(errors.New("can't use --copy-dest on a remote which doesn't support server side move or copy"))
```
Comment (ncw) [inline fs/operations/operations.go]: ...if so this text needs to be updated.

File: docs/content/docs.md [hunk 23]
```diff
@@ -475,6 +489,19 @@ The connection timeout is the amount of time rclone will wait for a
...
 `1m` by default.
 
+### --copy-dest=DIR ###
+
+When using `sync`, `copy` or `move` DIR is checked in addition to the 
+destination for files. If a file identical to the source is found that 
+file is server side copied from DIR to the destination. This is useful 
+for incremental backup.
+
+The remote in use must support server side move or copy and you must
```
Comment (ncw) [inline docs/content/docs.md]: I think this should say the remote must support server side copy not move.

**Comment by yparitcher** [comment]
you are right, I didn't realize some remotes only support one, will fix.

**Comment by yparitcher** [comment]
Now the test doesn't run on the `local` remote as it does not support copy only move

**Comment by ncw** [comment]
> Now the test doesn't run on the `local` remote as it does not support copy only move

Hmm, that is a little unfortunate but not the end of the world...

You'll need to add a test for Features().Copy and Skip the test if it isn't found to get the tests to pass.

I'm going to be away on vacation for a while, so communication will be sporadic !

**Comment by yparitcher** [comment]
should be ready

**Comment by ncw** [comment]
Apologies for the delay in responding, I've been on vacation.

That looks good now - thank you very much for implementing this very useful feature.

Will merge now :-)
